### ISSUE TITLE
RTL Performance Impact

### ISSUE THREAD

**Issue by WesleyHovick** [theKashey/react-focus-lock#125]
Greetings!

I've been investigating React Testing Library performance issues on a test suite in my organization. After digging I finally traced the issue to `<FocusLock>`. We are wrapping our modal with this and when I remove `<FocusLock>` from our modal component, it nearly cuts my testing time in half from 110s (83 tests total) to 50s!

I haven't had the time to dive into this repo so unfortunately I don't know where to begin to fix it, so I thought I'd open an issue.

What information can I give that will help? Is this a known performance issue?

**Comment by theKashey** [comment]
The issue is not known. To be more concrete I never checked how slow or fast it is.
The older versions([focus lock](https://github.com/theKashey/focus-lock/commit/5bc1498b6a7885d9e6ce906777395c31eca2bdef) <`0.8.0` has a `O(n^2)` in a one place, and that code is used in RTL to bring focus into.

I'll take a look. There are 2 possible origins of an error with two possible optimisations:
- is now "perf", it's "timeouts" - sidecar/effects adds a few _promises ticks_ to operations. It could delay overall system speed if many operations are in flight (read - you use jest)
- it's perf, but the "first render" can be (and actually is) optimized to auto focus one particular node, not trying to find the right one - https://github.com/theKashey/focus-lock/blob/master/src/focusMerge.ts#L43 - right now complexity of the operation is almost "flat", no matter what you do.

**Comment by WesleyHovick** [comment]
@theKashey Yep I am using jest!

After some 😴  I had the idea that perhaps I can `mock` the `<FocusLock>` components and just override any functions that might be taking a significant amount of time. What are your thoughts on that?

**Comment by theKashey** [comment]
You know, I personally in love with mocking of almost any sort. However, I am still interested in the details of your case:
- how many dom elements are on the page?
- are modals closed or opened? (is focus lock active or not?)
- and so on

I need some details so I can understand and represent the issue.

**Comment by WesleyHovick** [comment]
@theKashey I will try and get this info to you sometime today, looking like a really busy day though so it could be later.

**Comment by WesleyHovick** [comment]
We are opening just one modal at a time. This particular modal always has `<FocusLock>` wrapped around it. We do not provide an additional config or props to the `<FocusLock>` component, but I'm not sure if that means it's *active* or not. The general structure is similar to this.
```
<FocusLock>
     <StyledComponentWrapper>
          <ModalOverlay />
          <ModalContentWrapper>
               {props.children}
          </ModalContentWrapper>
     </StyledComponentWrapper>
</FocusLock>
```

This modal is used through our organization and is a shared custom built component.

In my particular case for the tests I'm running we are rendering a fairly simple modal. The DOM elements inside of the modal for the testing is somewhere between 103 and 111. It's essential a date/time picker, days of the week and 5 time of day options with a select button for each of the 5.

I've never done DOM counting before so I think those numbers are accurate. I added an id to the modal and ran the following in chrome console: `document.getElementById('timeSlotModal').getElementsByTagName('*').length`

Hope this helps! Let me know if I can provide any other info. Still working on getting mocking to work, not having a *ton* of success with it yet.

**Comment by theKashey** [comment]
> I've never done DOM counting before

FocusLocks analyses all nodes between `activeElement` and the `Lock`. On the initial render, the `activeElement` is body thus it will walk through all nodes, and if you use 2 months old focus-lock - that's `O(n^2)`

Could you please try to add `autoFocus` on any _focusable_ element inside a modal (a11y recommends focusing the modal itself, not the action button or first input) so that would disable the logic at all - focus lock does nothing if focus is already within the boundaries.

**Comment by WesleyHovick** [comment]
> if you use 2 months old focus-lock - that's O(n^2)

You know... I honestly hadn't even thought to check our version. We are at `"react-focus-lock": "1.18.3"` 😲  🤦‍♂️

**Comment by WesleyHovick** [comment]
Okay so I added `autoFocus` to the exit button for our modal and it brought the testing duration down to 72s from 114s. Quite interesting. Definitely not the 50s, but better for sure!

**Comment by theKashey** [comment]
> being on version < 0.8.0 

of `focus-lock`, not `react-focus-lock`. They are not actually bound to each other - you can have up to date `focus-lock` with the old `react-focus-lock`. Check your lock file.

> brought the testing duration down to 72s from 114s

Perfect sounds like a possibility of a quick win. You can downgrade back to `1.18.3` and check speed with it. Versions prior v2 might be a little faster - related to #122

**Comment by WesleyHovick** [comment]
@theKashey 
How will adding the `autoFocus` prop impact the use of `<AutoFocusInside>`. We have this as an optional prop for our custom component to enable this `<AutoFocusInside>{children}</AutoFocusInside>`. If we add `autoFocus` to an element outside of this, will the `AutoFocusInside` still function?

**Comment by theKashey** [comment]
`AutoFocusInside` and `autofocus` have roughly the same meaning, but if focus is __already__ inside the lock - it will do nothing.
You might consider `MoveFocusInside`, which is actually a "command".

**Comment by nathanforce** [comment]
Same problem here. I'm using the Dialog from `@reach`.

https://github.com/reach/reach-ui/blob/51e163dcfc2b3654e5344b719398daa3775cab63/packages/dialog/src/index.tsx#L180-L205

**Comment by nicolasschabram** [comment]
In our case, we've started seeing a test-performance regression after upgrading our `create-react-app` application to [v4](https://github.com/facebook/create-react-app/releases/tag/v4.0.0). The impact on the entire test suite was quite dramatic, something on the order of 4 to 5 times slower. We're using React Testing Library as well. Judging from the profiler, most of that time is spent in `react-focus-lock`'s `activateTrap()` function: 15 seconds on our slowest test! That test is probably rendering a bit too much for its own good, but nothing crazy (some tables and a form in a dialog on top).

![image](https://user-images.githubusercontent.com/1027642/103758673-cb936500-5012-11eb-886b-5dad25c83f95.png)

We're consuming `react-focus-lock` via `@reach/dialog`, so we worked around the issue by disabling it in test environments like this:

```tsx
 <DialogOverlay dangerouslyBypassFocusLock={process.env.NODE_ENV === "test"}>
```

CRA v4 came with `jest` and `jsdom` updates. Maybe something about those new versions doesn't play well with `react-focus-lock`?

**Comment by theKashey** [comment]
It's a few lines of code to stop traversing a whole tree when no action is required. Should find time to do it...

**Comment by theKashey** [comment]
The new version, 2.5.1, has been released with the potential fix for this issue (#151)

You can or update focus-lock, or update underlying `focus-lock`(just wipe it from yarn.lock or package.lock) which brought the fix - https://github.com/theKashey/focus-lock/pull/24

**Comment by theKashey** [comment]
the issue is considered to be resolved.
